### ISSUE TITLE
ParquetCachedBatchSerializer's performance on partitioned dataframe is worse than a single partitioned dataframe

### ISSUE THREAD

**Issue by razajafri** [NVIDIA/spark-rapids#4580]
There is a sizeable difference in performance when caching RDD on a single partition compared to a distributed run.

Consider the following Dataframe with 50,000 rows and a single column with 2048 elements in each array

```
scala> dataDf.printSchema
root
 |-- feature: array (nullable = false)
 |    |-- element: double (containsNull = true)
```

When we write it out to cache and read back the count, the times are as follows with and without repartitioning it to a single partition.

```
scala> spark.time(dataDf.cache.count)
Time taken: 37003 ms                                                            
res4: Long = 50000

scala> spark.catalog.clearCache

scala> spark.time(dataDf.repartition(1).cache.count)
Time taken: 15361 ms                                                            
res6: Long = 50000
```

When we do a rough benchmarking, the times (just for writing cache) are as follows for each of the runs 

|12 partitions run (ms)| Single partition (ms) |
|---------------- | -----------|
|1799||
|1741||
|1751||
|1753||
|1754||
|1753||
|1753||
|1753||
|1751||
|1762||
|1764||
|1765||
|21099|2471|

As apparent each of the partitioned parquet files written by the ParquetCachedBatchSeriazlier are small and collectively take a long time while the file written by single partition is large (amalgamation of all the partitioned files) and is much faster to write. 

I am not sure if there is a way to get around this problem as this is a known problem that GPU favors crunching large data over smaller ones

This [script](https://github.com/NVIDIA/spark-rapids/files/7902698/pca_cache_test.txt) was used to generate the above data

**Comment by revans2** [comment]
This looks like classic sub-linear scaling of the GPU. You give it 12 times as much data and it only got 40% slower to do it. From my calculations it looks like we are decoding about 781 MiB of data.  That is small enough to fit into a single columnar batch. I think this is works as expected.  But if you want to take some traces or see if this still happens with more data that would be interesting.

**Comment by razajafri** [comment]
The table size I am not sure of but the encoded in-memory parquet file was in fact 781MiB as you calculated.

**Comment by razajafri** [comment]
Can/Should the RapidsPlugin be intelligent about noticing this and essentially add the `repartition(1)` to the DF before caching it? Would that mean the cache is not going to be "distributed" anymore to the other nodes?

**Comment by revans2** [comment]
I don't think it can be intelligent about it. Users can be smart about it and if we update the docs it could be helpful to an end user. This is because moving all of the data to fewer hosts is not going to be free. If we knew the size of the data, network speed and compression speed for the shuffle we might be able to guess at a good number of workers to repartition the data to in order to optimize read speeds, but we don't know the size, we don't know how many times the data is going to be read vs written so we would have to assume worst case that it is read only once, and even if we did we did know all of that we would be doing something the user didn't ask us to do, so we would not want it to be a default operation.  Perhaps the user knows that they are going to do some heavy computation with the data and split it up on purpose to make the load more even, and we just go and undo it for them.
